// @colony/purser-trezor/class.js
/* @flow */

import GenericWallet from '@colony/purser-core/genericWallet';
import { userInputValidator } from '@colony/purser-core/helpers';
import { warning } from '@colony/purser-core/utils';
import { DESCRIPTORS, REQUIRED_PROPS } from '@colony/purser-core/defaults';
import { TYPE_HARDWARE, SUBTYPE_TREZOR } from '@colony/purser-core/types';

import type {
  TransactionObjectType,
  GenericClassArgumentsType,
} from '@colony/purser-core/flowtypes';

import { signTransaction, signMessage, verifyMessage } from './staticMethods';

import { classInstance as messages } from './messages';
import { REQUIRED_PROPS as REQUIRED_TREZOR_PROPS } from './defaults';

const { WALLET_PROPS } = DESCRIPTORS;

export default class TrezorWallet extends GenericWallet {
  constructor(propObject: GenericClassArgumentsType) {
    super(propObject);
    Object.defineProperties(this, {
      /*
       * Set the actual type and subtype (overwrite the generic ones)
       */
      type: { value: TYPE_HARDWARE, ...WALLET_PROPS},
      subtype: { value: SUBTYPE_TREZOR, ...WALLET_PROPS},
      sign: {
        
        value: async (transactionObject: TransactionObjectType) => {
            let requiredSignProps: Array<String> =
              REQUIRED_TREZOR_PROPS.SIGN_TRANSACTION;
            const { chainId = this.chainId, to } = transactionObject || {};
            /*
             * If we don't have a destination address, it means the user wants
             * to deploy a contract.
             *
             * For this the Trezor service requires a `inputData` value set.
             *
             * Otherwise, a `to` address *must* be set.
             */
            if (!to) {
              requiredSignProps =
                REQUIRED_TREZOR_PROPS.SIGN_TRANSACTION_CONTRACT;
              /*
               * Warn the user (in dev mode, at least) about Trezor's contract
               * deployment requirements
               */
              warning(messages.signContractDeployment);
            }
            /*
             * Validate the trasaction's object input
             */
            userInputValidator({
              firstArgument: transactionObject,
              requiredAll: requiredSignProps,
            });
            return signTransaction(
              { ...transactionObject, derivationPath: await this.derivationPath,
                chainId,},
            );
          },
        ...WALLET_PROPS,
      },
      signMessage: {
        
        value: async (messageObject: Object = {}) => {
            /*
             * Validate the trasaction's object input
             */
            userInputValidator({
              firstArgument: messageObject,
              requiredOr: REQUIRED_PROPS.SIGN_MESSAGE,
            });
            return signMessage({
              derivationPath: await this.derivationPath,
              message: messageObject.message,
              messageData: messageObject.messageData,
            });
          },
        ...WALLET_PROPS,
      },
      verifyMessage: {
        
        value: async (signatureVerificationObject: Object = {}) => {
            /*
             * Validate the trasaction's object input
             */
            userInputValidator({
              firstArgument: signatureVerificationObject,
              requiredAll: REQUIRED_PROPS.VERIFY_MESSAGE,
            });
            const { message, signature } = signatureVerificationObject;
            return verifyMessage({
              address: this.address,
              message,
              signature,
            });
          },
        ...WALLET_PROPS,
      },
    });
  }
}
